pkg/gardenlet/controller/managedseed/managedseed_reconciler.go

```diff
@@ -0,0 +1,230 @@
+		}
+		r.recordAndLog(managedSeed, fmt.Sprintf("Could not get shoot %s/%s: %+v", managedSeed.Namespace, managedSeed.Spec.Shoot.Name, err), condition,
+			gardencorev1beta1.ConditionFalse, "CouldNotGetShoot", corev1.EventTypeWarning, gardencorev1beta1.EventReconcileError, logger, logrus.ErrorLevel)
+		return nil, fmt.Errorf("could not get shoot %s/%s: %w", managedSeed.Namespace, managedSeed.Spec.Shoot.Name, err)
+	}
+	r.recordAndLog(managedSeed, fmt.Sprintf("Shoot %s found", kutil.ObjectName(shoot)), condition,
+		gardencorev1beta1.ConditionTrue, "ShootFound", corev1.EventTypeNormal, gardencorev1beta1.EventReconciling, logger, logrus.InfoLevel)
+
+	return shoot, nil
+}
+
+func (r *reconciler) checkSeedBootstrapped(ctx context.Context, managedSeed *seedmanagementv1alpha1.ManagedSeed, condition *gardencorev1beta1.Condition, logger *logrus.Entry) (bool, error) {
+	// Get seed
+	seed := &gardencorev1beta1.Seed{}
+	if err := r.gardenClient.DirectClient().Get(ctx, kutil.Key(managedSeed.Name), seed); err != nil {
+		if apierrors.IsNotFound(err) {
+			r.recordAndLog(managedSeed, fmt.Sprintf("Seed %s not found", managedSeed.Name), condition,
+				gardencorev1beta1.ConditionFalse, "SeedNotFound", corev1.EventTypeNormal, gardencorev1beta1.EventReconciling, logger, logrus.InfoLevel)
+			return false, nil
+		}
+		r.recordAndLog(managedSeed, fmt.Sprintf("Could not get seed %s", managedSeed.Name), condition,
+			gardencorev1beta1.ConditionFalse, "CouldNotGetSeed", corev1.EventTypeWarning, gardencorev1beta1.EventReconcileError, logger, logrus.ErrorLevel)
+		return false, fmt.Errorf("could not get seed %s: %w", managedSeed.Name, err)
+	}
+
+	// Get seed condition
+	sc := gardencorev1beta1helper.GetCondition(seed.Status.Conditions, gardencorev1beta1.SeedBootstrapped)
+	if sc == nil {
+		return false, nil
+	}
+
+	if sc.Status != gardencorev1beta1.ConditionTrue {
+		r.recordAndLog(managedSeed, fmt.Sprintf("Seed %s not bootstrapped: %s", managedSeed.Name, sc.Message), condition,
+			sc.Status, "SeedNotBootstrapped", corev1.EventTypeNormal, gardencorev1beta1.EventReconciling, logger, logrus.InfoLevel)
+		return false, nil
+	}
+	r.recordAndLog(managedSeed, fmt.Sprintf("Seed %s bootstrapped: %s", managedSeed.Name, sc.Message), condition,
+		sc.Status, "SeedBootstrapped", corev1.EventTypeNormal, gardencorev1beta1.EventReconciled, logger, logrus.InfoLevel)
+	return true, nil
+}
+
+func (r *reconciler) recordAndLog(managedSeed *seedmanagementv1alpha1.ManagedSeed, message string, condition *gardencorev1beta1.Condition, conditionStatus gardencorev1beta1.ConditionStatus, conditionReason, eventType, eventReason string, logger *logrus.Entry, logLevel logrus.Level) {
+	*condition = gardencorev1beta1helper.UpdatedCondition(*condition, conditionStatus, conditionReason, message)
+	logger.Log(logLevel, message)
+	r.recorder.Eventf(managedSeed, eventType, eventReason, "%s", message)
```

Comment:
Much of what you didn't like is due to waiting for the seed to become bootstrapped, since this causes the managed seed to be reconciled multiple times. I didn't like it either. Once it's removed, it will become much more straightforward.

I think we need an event when:
* The `ShootReconciled` condition is updated (either still reconciling or already reconciled), since shoot reconciliation takes some time and an event makes it more clear what's happening during this time.
* The `SeedRegistered` condition is updated (either success or error), since this is the main purpose of the controller. Note that it's impossible (or at least rather hard) to distinguish between the first time it happens and next times, so an event is fired always in case of success.

As for the `ShootExists` condition, I would propose to have an event only in the case of errors. We ensure via other means that the shoot exists, so we don't need an event to confirm this in case of success.

Reply:
Done as proposed above.